zokrates_core/src/flatten/mod.rs

```diff
@@ -538,6 +538,11 @@ impl<'ast, T: Field> Flattener<'ast, T> {
 
                 FlatExpression::Identifier(sub_bits_be[bitwidth - 1])
             }
+            BooleanExpression::EqBool(box lhs, box rhs) => self.flatten_boolean_expression(
+                symbols,
+                statements_flattened,
+                BooleanExpression::And(box lhs.clone(), box rhs.clone()),
```

Comment:
Isn't this incorrect? `false && false == false` but `(false == false) == true`
Also the cloning sounds unnecessary here.

One ok way to do this would be to use the same flattening as for `BooleanExpression::Eq` below, though with Booleans we have the guarantee that they flatten to either 0 or 1, so maybe this extra guarantee makes a cheaper check possible. The Field one is 2 constraints right now so it would need to be 1 constraint to be better I suppose?

Reply:
I agree, wasn't thought through enough from my side. 

 ```
    BooleanExpression::BoolEq(box lhs, box rhs) => self.flatten_boolean_expression(
                symbols,
                statements_flattened,
                BooleanExpression::Or(
                    box BooleanExpression::And(box lhs.clone(), box rhs.clone()), 
                    box BooleanExpression::And(
                        box BooleanExpression::Not(box lhs), 
                        box BooleanExpression::Not(box rhs)
                    ),
                )
            ),
```
Something like this should work functionally right, though I'm not sure how efficient this is. I don't think I understand the flattening part of this comment. What do you mean by the Field one?

Reply:
I meant "equality for field elements" as implemented in `flatten.rs` for `BooleanExpression::Eq`.

> I'm not sure how efficient this is

You could give it a try with a simple program:
```
def main(bool a, bool b) -> (bool):
   return a == b
```